### ISSUE TITLE
Auto-restart makes it less obvious when parallel jobs are failing

### ISSUE THREAD

**Issue by nspope** [popgenmethods/SINGER#9]
SINGER will automatically restart if something goes wrong. This can make it hard to determine if something is going wrong if running SINGER on a ton of chunks in parallel, because restarted chunks will overwrite the same log files. For these situations, it would be nice to have an option to disable the auto-debug, so that it's immediately obvious where things are going wrong (without monitoring stdout) and job managers like SLURM and snakemake don't waste compute.

Thanks for the fantastic tool!

**Comment by YunDeng98** [comment]
Hi @nspope, thanks for the information! I would like to know a bit more details before I start working on it: 

(1) Is this using `parallel_singer` or some parallelization of your own implementation? If the former, I should work on fixing `parallel_singer` too;

(2) When you say "restarted chunks will overwrite the same log files", could you describe a bit more about its behavior? Is it like SINGER will start threading everything from the start? Or is it rather move backwards for a few iterations and restart from there before failing again?

Thanks a lot!

**Comment by nspope** [comment]
Maybe an alternative to disabling auto-restart entirely would be an option to set the maximum number of attempts to something lower than 100?

**Comment by YunDeng98** [comment]
OK weird behavior in (2) and that certainly shouldn't happen. I will write a new `parallel_singer` with an option to turn it off. So the reason why it auto re-starts is because there are some extremely infrequency numerical errors which might happen one in a millions threadings. I haven't fully figured all of them out, but it is annoying enough that if we do thousands of MCMC iterations for thousands of chunks of the entire genome some bugs would occur. What would you recommend as a best option? Do you want to a new script which you can controllably use to resume partially-completed runs?

**Comment by YunDeng98** [comment]
but restarting from beginning every time is certainly not what I intended, and that is not helpful at all.

**Comment by nspope** [comment]
Ah, that makes sense -- I would suggest:

(1). Append to rather than overwrite the log file for each chunk if something goes wrong, so it's easier to determine than sifting through stdout from the entire parallel run

(2). Make the max number of restarts a settable option in singer_master / singer_parallel

**Comment by YunDeng98** [comment]
Thanks a lot, Nate! These are all very sound advice. I feel like the python wrapping needs to be more user-friendly/user-controlled. My experience and other feedbacks said that if the auto-restart acts on MCMC it works great, but opposite when it restarts from threading initialization. In nutshell, the MCMC re-boot works well, because it means the program mostly runs and is only interrupted by minor bugs. The real problem is when something is very wrong and the program would spend 100 times trying non-meaningful stuff before bugging out, and people don't see anything on the log. I will start from there by adding stricter control of initial phase of threading (before MCMC).

**Comment by nspope** [comment]
hmm, I think I'm seeing chunks restart from the initial threading after a good amount of progress -- e.g. after 2000 MCMC iterations beyond the initial thread. So this definitely doesn't seem like the desired behavior.

In case it's helpful to figure out why the algo isn't restarting from the last iteration, the assertions getting tripped are one of:
```
singer: Branch.cpp:16: Branch::Branch(Node_ptr, Node_ptr): Assertion `(l == nullptr and u == nullptr) or l->time < u->time' failed.
singer: approx_BSP.cpp:159: void approx_BSP::null_emit(double, Node_ptr): Assertion `ws > 0' failed.
singer: Polar_emission.cpp:86: virtual double Polar_emission::mut_emit(Branch&, double, double, double, std::set<double>&, Node_ptr): Assertion `emit_prob > 0' failed.
```
(generally speaking it seems I can avoid these by using shorter chunks)

**Comment by YunDeng98** [comment]
Hi @nspope, would it be possible for you to send me a 1Mb chunk to replicate this behavior?

**Comment by nspope** [comment]
Sure, let me verify that I can replicate outside of the parallel wrapper with a fixed seed, and I'll ping you over slack.
